### PR TITLE
Changes from #297

### PR DESCRIPTION
**Pull request by KateMoreva**
Changes from https://github.com/mobie/mobie-viewer-fiji/pull/297

### CODE REVIEW

**Review by constantinpape — APPROVED** [review]
Looks good to me. Feel free to merge your on PRs in this repository if you think they are ready to go @KateMoreva.
Since we are not depending on this yet in MoBIE there's no risk of breaking things :).

**Comment by KateMoreva** [comment]
> Looks good to me. Feel free to merge your own PRs in this repository if you think they are ready to go @KateMoreva.
> Since we are not depending on this yet in MoBIE there's no risk of breaking things :).

Thanks, I started working on cleaning the code as it a bit essential to work with 2-5d, once it is done (hope today) I'll merge.

**Comment by KateMoreva** [comment]
- dimensionSeparator is read from metadata
- can open example ome.zar files of both versions from file system (form s3 fully not tested yet (can open https://s3.embl.de/i2k-2020/em-raw.ome.zarr))
In progress:
[ ] - functionality should be tested for different scales (possible problems with 2d and 4d)
[ ] - code cleaning (also for n5-zarr as a dependency)

**Comment by KateMoreva** [comment]
I was working on opening 2d and 4d ome.zarr images, but faces some errors. @tischi, if you ever faced such problems, can you please share ideas on how to solve them?

When I try to open 4d image I get:
```
Exception in thread "PainterThread" java.lang.ArrayIndexOutOfBoundsException: 3
	at net.imglib2.img.cell.CellGrid.cellDimensions(CellGrid.java:153)
	at bdv.viewer.render.ProjectorFactory.prefetch(ProjectorFactory.java:317)
	at bdv.viewer.render.ProjectorFactory.createSingleSourceVolatileProjector(ProjectorFactory.java:241)
	at bdv.viewer.render.ProjectorFactory.createSingleSourceProjector(ProjectorFactory.java:202)
	at bdv.viewer.render.ProjectorFactory.createProjector(ProjectorFactory.java:170)
	at bdv.viewer.render.MultiResolutionRenderer.createProjector(MultiResolutionRenderer.java:628)
	at bdv.viewer.render.MultiResolutionRenderer.paintFullFrame(MultiResolutionRenderer.java:465)
	at bdv.viewer.render.MultiResolutionRenderer.paint(MultiResolutionRenderer.java:439)
	at bdv.viewer.ViewerPanel.paint(ViewerPanel.java:473)
	at bdv.viewer.render.PainterThread.run(PainterThread.java:82)
```

**Comment by tischi** [comment]
> When I try to open 4d image I get

Could you please put code into `src/test` that opens such an image?
Then I can have a look!

Just one comment already: I think that the output that you feed into SpimData **must** have the same number of dimensions as in the old code where we were reading 5D data. That means if you have now less dimensions somewhere you may have to  add singleton dimensions using `Views.addDimension( )`;

**Comment by KateMoreva** [comment]
> > When I try to open 4d image I get
> 
> Could you please put code into `src/test` that opens such an image?
> Then I can have a look!
> 
> Just one comment already: I think that the output that you feed into SpimData **must** have the same number of dimensions as in the old code where we were reading 5D data. That means if you have now less dimensions somewhere you may have to add singleton dimensions using `Views.addDimension( )`;

Thank you! I'll try to add a new dimension and load the code if the problem still exists.

**Comment by KateMoreva** [comment]
The reader can open files with cyx, czyx, tcyx, tczyx, tyx, tzyx, zyx axes and old versions without axes in .zattrs

**Comment by constantinpape** [comment]
So it indeed seems like supporting 2d data is a bit more complex, see https://github.com/mobie/mobie-viewer-fiji/issues/310. I think we should go ahead and merge this when everything but 2d works and then work on the 2d problem separately.

**Comment by tischi** [comment]
ok

File: src/main/java/de/embl/cba/bdv/ome/zarr/loaders/N5OMEZarrImageLoader.java
Comment (constantinpape) [inline src/main/java/de/embl/cba/bdv/ome/zarr/loaders/N5OMEZarrImageLoader.java:152]: if we have 5 axes we know that axes are "tczyx", so you only need the first "if", not the second "else if". (The conditions are identical.)

File: src/main/java/de/embl/cba/bdv/ome/zarr/loaders/N5OMEZarrImageLoader.java
Comment (constantinpape) [inline src/main/java/de/embl/cba/bdv/ome/zarr/loaders/N5OMEZarrImageLoader.java:160]: This coundition and the next one (line 164) are identical in this context and it should be enough to check for 
`nDim == 4 && axesMap.containsKey("t")`

File: src/main/java/de/embl/cba/bdv/ome/zarr/zarr/N5ZarrReader.java
Comment (constantinpape) [inline src/main/java/de/embl/cba/bdv/ome/zarr/zarr/N5ZarrReader.java:58]: Some general comments on this class and the overall design:
- This class includes logic that is special to ome.zarr, e.g. the `axesMap`, hence I think it should be called `N5OmeZarrReader`.
- What's the relation to the `OmeZarrReader`? Do we really need to have both classes?
- We should somewhere check that the `axes` are valid, i.e. that they are in one of the combinations that we support, e.g. `czyx`, `zyx`, etc. (Note that the ome.zarr spec is currently not very clear if something like `xyzc` could exist, so we should have explicit checks for axes combinations that are not supported by us).

**Review by constantinpape — COMMENTED** [review]
The changes look very good overall. I left a few comments on where I think we could improve a few things.

File: src/main/java/de/embl/cba/bdv/ome/zarr/zarr/N5ZarrReader.java
Comment (KateMoreva) [reply src/main/java/de/embl/cba/bdv/ome/zarr/zarr/N5ZarrReader.java]: - I re named the class
- As for the code duplication:
`OmeZarrReader` is an analog to the `OMEZarrS3Reader`, but non-s3 one is for the file system.
`N5S3ZarrReader` has a lot in common with `N5OmeZarrReader` 
`N5ImageLoader` has a lot in common with `N5OMEZarrImageLoader`
But I think that we slowly are solving these problemms.

Comment (constantinpape) [reply src/main/java/de/embl/cba/bdv/ome/zarr/zarr/N5ZarrReader.java]: Thanks for the clarification re the second point. Let's keep it separate for now and see if it's possible to merge it later.

**Review by constantinpape — APPROVED** [review]
This looks good to me now. Feel free to merge whenever you think it's good to go.

**Pull request merged**
